extension/js/common/api/shared/api.ts

```diff
@@ -60,37 +83,152 @@ export class Api {
+    const newTimeoutPromise = (): Promise<never> => {
+      return new Promise((_resolve, reject) => {
+        /* error-handled */ setTimeout(() => {
+          reject(AjaxErr.fromXhr({ readyState, status: -1, statusText: 'timeout' }, reqContext)); // Reject the promise with a timeout error
+        }, req.timeout ?? 20000);
+      });
+    };
+    let body: BodyInit | undefined;
+    let duplex: 'half' | undefined;
+    let uploadPromise: () => void | Promise<void> = Value.noop;
+    let url: string;
+    if (req.method === 'GET' || req.method === 'DELETE') {
+      if (typeof req.data === 'undefined') {
+        url = req.url;
+      } else {
+        url = Url.create(req.url, req.data, false);
+      }
+    } else {
+      url = req.url;
+      if (req.method === 'PUT' || req.method === 'POST') {
+        if ('data' in req && typeof req.data !== 'undefined') {
+          if (req.dataType === 'JSON') {
+            body = JSON.stringify(req.data);
+            headersInit.push(['Content-Type', 'application/json; charset=UTF-8']);
+          } else if (req.dataType === 'TEXT') {
+            if (supportsRequestStreams && req.progress?.upload) {
+              const upload = req.progress?.upload;
+              const transformStream = new TransformStream();
+              uploadPromise = async () => {
+                const transformWriter = transformStream.writable.getWriter();
+                for (let offset = 0; offset < req.data.length; ) {
+                  const chunkSize = Math.min(1000, req.data.length - offset);
+                  await Promise.race([transformWriter.write(Buf.fromRawBytesStr(req.data, offset, offset + chunkSize)), newTimeoutPromise()]);
+                  upload((offset / req.data.length) * 100, offset, req.data.length);
+                  offset += chunkSize;
+                }
+                await Promise.race([transformWriter.close(), newTimeoutPromise()]);
+              };
+              body = transformStream.readable;
+              duplex = 'half'; // activate upload progress mode
             } else {
-              resolve(data as unknown);
+              body = req.data;
             }
-          })
-          .catch(reject);
-      });
+            if (typeof req.contentType === 'string') {
+              headersInit.push(['Content-Type', req.contentType]);
+            }
+          } else {
+            body = req.data; // todo: form data content-type?
+          }
+        }
+      }
+    }
+    const abortController = new AbortController();
+    const requestInit: RequestInit & { duplex?: 'half' } = {
+      method: req.method,
+      headers: headersInit,
+      body,
+      duplex,
+      mode: 'cors',
+      signal: abortController.signal,
+    };
+    let readyState = 1; // OPENED
+    const reqContext = { url: req.url, method: req.method, data: body, stack: req.stack };
+    try {
+      const fetchPromise = fetch(url, requestInit);
```

Comment:
what's the benefit of `fetch()` over `$.ajax` ?

Reply:
This was done per this migration instruction [Replace XMLHttpRequest() with global fetch()](https://developer.chrome.com/docs/extensions/migrating/to-service-workers/#replace-xmlhttprequest)

Reply:
huh. i see. there could be a small library available that encapsulates the more complicated parts of the fetch handling? perhaps there's a bug there, and it could simplify it.